test/unit/properties.test.ts

```diff
@@ -0,0 +1,425 @@
+/*
+ * sonar-scanner-npm
+ * Copyright (C) 2022-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import { DEFAULT_SONAR_EXCLUSIONS, SCANNER_BOOTSTRAPPER_NAME } from '../../src/constants';
+import { LogLevel, log } from '../../src/logging';
+import { FakeProjectMock } from './mocks/FakeProjectMock';
+
+jest.mock('../../src/logging');
+
+const projectHandler = new FakeProjectMock();
+
+describe('getProperties', () => {
+  describe('should handle JS API scan options params correctly', () => {
+    it('should detect and use user-provided scan option params', () => {
+      projectHandler.reset('fake_project_with_sonar_properties_file');
+
+      const properties = projectHandler.run({
+        serverUrl: 'http://localhost/sonarqube',
+        token: 'dummy-token',
+        verbose: true,
+        options: {
+          'sonar.projectKey': 'use-this-project-key',
+        },
+      });
+
+      expect(properties).toEqual({
+        ...projectHandler.getExpectedProperties(),
+        'sonar.host.url': 'http://localhost/sonarqube',
+        'sonar.token': 'dummy-token',
+        'sonar.verbose': 'true',
+        'sonar.projectKey': 'use-this-project-key',
+        'sonar.projectName': 'Foo',
+        'sonar.projectVersion': '1.0-SNAPSHOT',
+        'sonar.sources': 'the-sources',
+      });
+    });
+
+    it('should not set verbose mode when explicitly turned off', () => {
+      projectHandler.reset('fake_project_with_sonar_properties_file');
+
+      const properties = projectHandler.run({
+        serverUrl: 'http://localhost/sonarqube',
+        verbose: false,
+      });
+
+      expect(properties['sonar.verbose']).toBe('false');
+    });
+  });
+
+  describe('should handle package.json correctly', () => {
+    it('should generate default properties with package.json', () => {
```

Comment:
a handful of tests fail for me with a token defined when it isn't expected

Reply:
Any change you'd have `SONAR_TOKEN` defined in your global environment?

Reply:
looks much better, still have one test failing with the string format

```
  Expected: "WARN", "Failed to parse JSON parameters from ENV: SyntaxError: Unexpected token h in JSON at position 1"
    Received: "WARN", "Failed to parse JSON parameters from ENV: SyntaxError: Unexpected token 'h', \"this is def not JSON\" is not valid JSON"
     ```
     
     Might need to wait until we align on the logger

Reply:
Damn, looks like MacOS and Linux error message for JSON parsing are different 😅 Should be fixed now

Reply:
looks good